compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1130,8 +1130,18 @@ fn should_encode_trait_impl_trait_tys(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
 }
 
 // Return `false` to avoid encoding impl trait in trait, while we don't use the query.
-fn should_encode_fn_impl_trait_in_trait<'tcx>(_tcx: TyCtxt<'tcx>, _def_id: DefId) -> bool {
-    false
+fn should_encode_fn_impl_trait_in_trait<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
+    if tcx.sess.opts.unstable_opts.new_rpitit {
+        match tcx.def_kind(def_id) {
+            DefKind::Fn | DefKind::AssocFn => {
+                tcx.def_kind(tcx.parent(def_id)) == DefKind::Trait
+                    && tcx.hir().get_fn_output(def_id.expect_local()).is_some()
```

Comment:
Should we encode only if we know the trait has RPITITs? Is that easy to check?

Reply:
Well, I guess we can `associated_items_for_impl_trait_in_trait` query and check if not empty?.